designs/error-reporting-design.md

```diff
@@ -0,0 +1,341 @@
+<!-- # TODOs:
+- Review everything in "Examples of Design"
+- Review "Changes to Data Structure".
+- Review "Functions to add"
+- Review "Future Work" -->
+
+# Error Reporting
+
+## Design
+### What we aim to achieve by next week
+1. More information in all error messages
+    1. [Report Line Number](#report-line-number)
+    2. [Report Column Number](#report-column-number)
+    3. [Report Source Code](#report-source-code)
+    4. [Pretty Source Code](#pretty-source-code)
+
+2. Better type error
+    1. [Assignment with different types](#Assignment-with-different-types)
+    2. [Override \_\_init\_\_ signature](#override-\_\_init\_\_-signature)
+    3. [Binary operator type hint](#binary-operator-type-hint)
+    4. [Condition expression type hint](#condition-expression-type-hint)
+    5. [Type check print/len argument](#type-check-print/len-argument)
+
+3. Better parse error
+    1. [Using keywords as variable names](#using-keywords-as-variable)
+
+4. Better Run-time error
+    1. [Divide by zero](#divide-by-zero)
+
+### Future Work
+1. [Stack trace](#stack-trace)
+
+2. Better type error
+    1. [Function return check](#function-return-check)
+    2. [Branch return hint](#branch-return-hint)
+
+2. Better Run-time error
+    1. [Assert not none](#assert-not-none)
+    2. [Out of memory](#out-of-memory)
+
+
+
+## Changes to Data Structures
+
+The only changes we would add to the AST is the following:
+
+- We're extending the annotation type `A` to `Annotation` with information of start and end location (line and column numbers) as well as the full line of source code that contains the given component.
+- Such information except for the type information will be added during parsing, and type information will be added after type checking.
+
+Here's an overview of the changes in code:
+
+```ts
+Annotation = {
+    type?: Type,
+    fromLoc: Loc,
+    toLoc: Loc,
+    fullSrc: string
+}
+
+Loc = {
+    line: num,
+    col: num,
+}
+```
+
+## Functions to Add
+### Parsing
+During parsing, we will need to record the starting and ending location of each AST node. We need to figure out how to get line and column number out of lezer or count them ourselves. We may add `annotateLocation(ast, lezer_cursor)` function that calculates location information from a lezer cursor and add to the annotation field of the AST node.
+
+### Type Checking
+We will add a collection of functions, such as `binaryOperatorTypeError(ast)` that helps format the error message using the annotation in the AST node.
+
+### Compiler
```

Comment:
Oh, cool! If you do this, talk to the memory mangement group. They might want this.